arch/xtensa/src/esp32/esp32_ble_adapter.c

```diff
@@ -0,0 +1,2814 @@
+
+  for (int i = 0; i < 32; i++)
+    {
+      bit = 1 << i;
+      if (bit & mask)
+      {
+        wlinfo("Enabled bit %d\n", i);
+        up_enable_irq(i);
+      }
+    }
+}
+
+static bool is_wifi_clk_peripheral(periph_module_e periph)
+{
+  /* A small subset of peripherals use WIFI_CLK_EN_REG and
+   * CORE_RST_EN_REG for their clock & reset registers
+   */
+
+  switch (periph)
+  {
+    case PERIPH_SDMMC_MODULE:
+    case PERIPH_SDIO_SLAVE_MODULE:
+    case PERIPH_EMAC_MODULE:
+    case PERIPH_RNG_MODULE:
+    case PERIPH_WIFI_MODULE:
+    case PERIPH_BT_MODULE:
+    case PERIPH_WIFI_BT_COMMON_MODULE:
+    case PERIPH_BT_BASEBAND_MODULE:
+    case PERIPH_BT_LC_MODULE:
+        return true;
+    default:
+        return false;
+  }
+}
+
+static uint32_t get_clk_en_mask(periph_module_e periph)
+{
+  switch (periph)
+  {
+    case PERIPH_BT_MODULE:
+      return DPORT_WIFI_CLK_BT_EN_M;
+    default:
+      return 0;
+  }
+}
+
+static uint32_t get_rst_en_mask(periph_module_e periph, bool enable)
+{
+  switch (periph)
+  {
+    case PERIPH_WIFI_MODULE:
+    case PERIPH_BT_MODULE:
+    case PERIPH_WIFI_BT_COMMON_MODULE:
+    case PERIPH_BT_BASEBAND_MODULE:
+    case PERIPH_BT_LC_MODULE:
+      return 0;
+    default:
+      return 0;
+  }
+}
+
+static uint32_t get_clk_en_reg(periph_module_e periph)
+{
+  return is_wifi_clk_peripheral(periph) ? DPORT_WIFI_CLK_EN_REG :
+                                          DPORT_PERIP_CLK_EN_REG;
+}
+
+static uint32_t get_rst_en_reg(periph_module_e periph)
+{
+  return is_wifi_clk_peripheral(periph) ? DPORT_CORE_RST_EN_REG :
+                                          DPORT_PERIP_RST_EN_REG;
+}
+
+static void bt_periph_module_enable(periph_module_e periph)
+{
+  DPORT_SET_PERI_REG_MASK(get_clk_en_reg(periph),
+                          get_clk_en_mask(periph));
+  DPORT_CLEAR_PERI_REG_MASK(get_rst_en_reg(periph),
+                            get_rst_en_mask(periph, true));
+}
+
+/****************************************************************************
+ * Name: esp_int_adpt_cb
+ *
+ * Description:
+ *   BT interrupt adapter callback function
+ *
+ * Input Parameters:
+ *   arg - interrupt adapter private data
+ *
+ * Returned Value:
+ *   NuttX error code
+ *
+ ****************************************************************************/
+
+static int esp_int_adpt_cb(int irq, void *context, void *arg)
+{
+  struct irq_adpt_s *adapter = (struct irq_adpt_s *)arg;
+
+  adapter->func(adapter->arg);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: interrupt_disable
+ *
+ * Description:
+ *   Enter critical section by disabling interrupts and taking the spin lock
+ *   if in SMP mode.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void IRAM_ATTR interrupt_disable(void)
+{
+  g_inter_flags = enter_critical_section();
+}
```

Comment:
Alan, please talk to @Ouss4 and see if here it's not the case of using `spin_lock_irqsave`.
Besides that, what's the real purpose of this function because its name is `interrupt_disable` but the description is to enter critical section. If it's only to disable interrupts you may use `up_irq_save`

Reply:
> if it's only to disable interrupts you may use up_irq_save

Yes, if we only need to disable interrupts in the current CPU, `up_irq_save` should be used.

For replacing `enter_critical_section` with `spin_lock_irqsave` we have to verify that we are using the critical section to protect data only and not causing any context switching.  The issue here is that these functions go through the wireless library and we don't really know what's going on inside.  You should be able to track the calls between `interrupt_disable` and `interrupt_restore` with a debugger.